### PR TITLE
portable: extract polkit and D-Bus configuration files

### PR DESCRIPTION
**Pull request by bluca**
If a portable image ships polkit and/or D-Bus configuration files, extract them to the respective /etc/ directories, mark them with an xattr (so that they can be recognized for removal), and reload the respective service if needed (dbus only, as polkit watches with inotify already and doesn't support reloading).

This allows portable services to provide services over D-Bus, gated by polkit policies.

### CODE REVIEW

**Comment by poettering** [comment]
Sorry, but no, this is terrible. let's not mix and match stuff in dirs owned by other packages. This is terrible. And in particular let's not do this with explicit hookups for some select packages. That doesn't scale.

Sorry, but this is not thought to the end, and should not be done like this.

**Comment by bluca** [comment]
> let's not mix and match stuff in dirs owned by other packages

It doesn't. Those directories are not owned by other packages - it's not /usr, it's /etc, and it's intended to be used to deploy configuration _not_ owned by the package, but that comes from other sources, be it a local user or a config deployment system, that's the entire point of it. The xattr ensure we can figure out what's ours and what's not, so there's no risk of confusion - no more than there would be with any other configuration management system like etcd or ansible or those sort of things.

> And in particular let's not do this with explicit hookups for some select packages. That doesn't scale.

Yes, and it's not intended to. This is for integration in a carefully selected and basic set of components that normal system services need to integrate with. It's not intended for anything and everything, but just the basics that everybody need to integrate a low-level system service into Linux, and right now that's D-Bus and Polkit. This is the only possible way to do it, there are no other realistic solutions.

It's the single major drawback of portable services right now that this basic integration is impossible, and it really needs to be solved.

**Comment by poettering** [comment]
Sorry, but I am very strongly against this.

Make multi-modal DDIs a thing, it's a much better approach

**Comment by YHNdnzj** [comment]
Yeah, let's not step into directories owned by other softwares deliberately. Instead, when building the image the files should be installed to a well-defined location, where we'd pick them up later.

**Comment by poettering** [comment]
i.e. a simple approach could be to introduce a new field to a portable service's os-release file PORTABLE_APPLY_SYSEXT=0/1. if enabled we'd simply symlink the relevant portable service DDIs that are enabled into /run/confexts/… and then issue a confext refresh request via varlink.

And we'd have to define some flags in the gpt partition flags that indicate the intended scope of a root or usr partition. i.e. a set of flags for "use-as-sysext", "use-as-confext", "use-as-portable", "use-as-boot" or so.

a multi-modal DDI that contains a sysext and a portable service would then set "use-as-sysext" on one root partition and "use-as-portable" on the other.

Note that these flags are just to direct the dissector to the right partition, they are authenticated by the fact of extension-release and os-release files exist within these directories.

**Comment by bluca** [comment]
> Sorry, but I am very strongly against this.
> 
> Make multi-modal DDIs a thing, it's a much better approach

It really isn't, it cannot work for D-Bus and polkit. Both of them run in a private mount namespace due to sandboxing, so any "multi-modal DDI", by which I assume you mean some scheme where a sysext is used on the root filesystem, will simply be invisible to those services when attaching a portable image, as they are running off the bare rootfs, and won't see the new overlay. You would need to restart them so that they pick the new mounts from the rootfs, but then of course you cannot restart D-Bus as you well know, so you would need a reboot to attach/upgrade a portable image, which is bonkers of course. And for polkit a restart is needed too, as it watches /etc/ with inotify, but of course the inotify is on the inode on the original rootfs, so the new overlay is simply ignored, so a restart is needed there too.

And even if rebooting to install a portable image was acceptable (it's really not, I hope I don't have to explain why), it's a horrible scheme: installing a portable image, the point of which is to have a separate file system independent of the rootfs, goes in by the backdoor and messes up the rootfs with a new overlay? That would be absolutely dreadful

So this idea I'm afraid just doesn't work, and even if it worked it would be really bad.

On top of that, as you know I do not and cannot use GPT images, so even if it worked, it would be completely useless to me, so I would certainly not waste any time working on that, it would not make any sense, there's no way to justify spending that time. I mean of course if somebody else really likes that scheme and adds it as an alternative I'm not going to stand in the way, by all means go for it, but it for sure cannot be the only solution.

**Comment by bluca** [comment]
> Yeah, let's not step into directories owned by other softwares deliberately. Instead, when building the image the files should be installed to a well-defined location (could be another image), where we'd pick them up later.

/etc/ is not owned by other software, it is very much intended for configuration deployed NOT by that software (configuration owned by a piece of software goes in /usr/). Deploying config files to /etc is absolutely normal for any orchestration/configuration/etc system. And also I'm not sure what "we'd pick them up later" would mean? polkit config files need to be in /etc/polkit, dbus config files need to be in /etc/dbus-1 and so on - and they really have to be there, on the filesystem that the services are actually using, and cannot be symlinks either due to sandboxing.

**Comment by bluca** [comment]
And in fact I really do not understand where this idea that you cannot install config files into "other packages" directories comes from - it's literally how everything works. Thousands of softwares ship config files that are installed in /usr/lib/systemd/ and /usr/share/polkit and /usr/share/dbus-1 and so on. Heck polkit ships a config file installed in dbus-1 and one in systemd/system, and dbus ships one for systemd, and so on and so forth. A software installing configuration files in another package's directory is literally how things are integrated in Linux, including all of our stuff. If that rule was actually true, then it would mean the systemd package would be the only one allowed to ship units, and every single software on the planet would need to get their units merged in the systemd package in order to ship it. And all polkit rules would need to be merged in the polkit package. And so on. That's not really how any of this works?

**Comment by poettering** [comment]
> It really isn't, it cannot work for D-Bus and polkit. Both of them run in a private mount namespace due to sandboxing, so any "multi-modal DDI", by which I assume you mean some scheme where a sysext is used on the root filesystem, will simply be invisible to those services when attaching a portable image, as they are running off the bare rootfs, and won't see the new overlay. You would need to restart them so that they pick the new mounts from the rootfs, but then of course you cannot restart D-Bus as you well know, so you would need a reboot to attach/upgrade a portable image, which is bonkers of course. And for polkit a restart is needed too, as it watches /etc/ wit

Sandboxing is great, but it shouldnt constrain us for making things like this possible. Sandboxing is only best effort stuff, it shouldnt break our uses. If we really need a hole we punch a hole.

And this is fixable as well because we should support landlock for such policies to precisely deal with mount propagation issues that collide with the sandboxing.

Also note that we already have support for reloading services on activation.

**Comment by poettering** [comment]
> the point of which is to have a separate file system independent of the rootfs, goes in by the backdoor and messes up the rootfs with a new overlay? That would be absolutely dreadful

Nah, sysext are fantastic for this because they after all do *not* mess up anything. You can apply them and get rid of them, no harm done. U like anything that writes all kind of shit into arbitrary dirs in /etc.

**Comment by poettering** [comment]
> On top of that, as you know I do not and cannot use GPT images

Deeply unimpressed by that. In 2024 we are not going to add additional infra with this as a reason. Just fix your toolchain. I never heard a technical reason why you couldnt fix your toolchain

Sorry i am not going to move on any of this. Let's stop adding hacks specific to some corner case usecases like this. 

You know you always have the chance to fix this in the offending programs btw. I.e. teach dbus-broker to take policy at name bind time and done. Everybody already agreed to that, hence if you address that properly the whole issue goes away.

**Comment by bluca** [comment]
> > It really isn't, it cannot work for D-Bus and polkit. Both of them run in a private mount namespace due to sandboxing, so any "multi-modal DDI", by which I assume you mean some scheme where a sysext is used on the root filesystem, will simply be invisible to those services when attaching a portable image, as they are running off the bare rootfs, and won't see the new overlay. You would need to restart them so that they pick the new mounts from the rootfs, but then of course you cannot restart D-Bus as you well know, so you would need a reboot to attach/upgrade a portable image, which is bonkers of course. And for polkit a restart is needed too, as it watches /etc/ wit
> 
> Sandboxing is great, but it shouldnt constrain us for making things like this possible. Sandboxing is only best effort stuff, it shouldnt break our uses. If we really need a hole we punch a hole.
> 
> And this is fixable as well because we should support landlock for such policies to precisely deal with mount propagation issues that collide with the sandboxing.

Disabling all sandboxing just to enable deploying portable images does not sound like a good idea at all to me, and the various upstreams would never accept it. Which means the only way for this to work is for portabled to add system-level drop-ins that disable all sandboxing globally. That sounds terrible to me.

> Also note that we already have support for reloading services on activation.

I don't think so? Perhaps you are thinking about those horrible hacks that were added to force a daemon-reload? But that's not the same as reloading individual services. And not all services actually support that - polkit for example doesn't, it watches its configs with inotify.

**Comment by YHNdnzj** [comment]
> And in fact I really do not understand where this idea that you cannot install config files into "other packages" directories comes from - it's literally how everything works. Thousands of softwares ship config files that are installed in /usr/lib/systemd/ and /usr/share/polkit and /usr/share/dbus-1 and so on.

But systemd is not a package manager. And portable services are _not_ packages. They are conceptually isolated from host, which is completely opposite from what you're trying to do now, i.e. to directly copy those files to host rootfs.

**Comment by poettering** [comment]
And please stop playing games with the needs-discussion label. Dropping it doesnt resolve the fact that this approach is no acceptable in my eyes.

**Comment by bluca** [comment]
> > And in fact I really do not understand where this idea that you cannot install config files into "other packages" directories comes from - it's literally how everything works. Thousands of softwares ship config files that are installed in /usr/lib/systemd/ and /usr/share/polkit and /usr/share/dbus-1 and so on.
> 
> But systemd is not a package manager. And portable services are _not_ packages. They are conceptually isolated from host, which is completely opposite from what you're trying to do now, i.e. to directly copy those files to host rootfs.

Portabled is a configuration manager, like etcd or ansible. It already writes several things to /etc/ when attaching, that's literally how it works today. Changing the host's mount table under its feet is the opposite of isolation.

**Comment by bluca** [comment]
> > the point of which is to have a separate file system independent of the rootfs, goes in by the backdoor and messes up the rootfs with a new overlay? That would be absolutely dreadful
> 
> Nah, sysext are fantastic for this because they after all do _not_ mess up anything. You can apply them and get rid of them, no harm done.

They mess up the mount table on the host. Which is fine if that's what one _wants_ to do explicitly, and if everything supports it. It's pretty much only useful for things that are run to completion, like users running binaries or one-shot short-running services. It really doesn't work well for anything else, especially not dynamically. Requiring a reboot to install a new portable service is just not going to work.

> U like anything that writes all kind of shit into arbitrary dirs in /etc.

That is literally how portabled works, since day one.

**Comment by bluca** [comment]
> > On top of that, as you know I do not and cannot use GPT images
> 
> Deeply unimpressed by that. In 2024 we are not going to add additional infra with this as a reason. Just fix your toolchain. I never heard a technical reason why you couldnt fix your toolchain

I've told you many times, and you don't like the answer - that is fine of course, but you can't use that to tell me to invest a gigantic amount of time and resources to go work on something completely different, that I have literally no use for. Again if someone wants to try and invest time to implement such a scheme, by all means go for it, but it's just not going to be me.

> Sorry i am not going to move on any of this. Let's stop adding hacks specific to some corner case usecases like this.

I am really not sure this can be defined as a "corner case", given I have the _very_ strong suspicion I am pretty much the only one in the world that uses any of this for real, in production. I mean either that, or the implementation is absolutely perfect and has zero bugs, and given at this point I've authored a very large chunk of it, I have very serious reservations about the latter option...

> You know you always have the chance to fix this in the offending programs btw. I.e. teach dbus-broker to take policy at name bind time and done. Everybody already agreed to that, hence if you address that properly the whole issue goes away.

No, that doesn't work, and it's not enough. It wouldn't work on reboot for example, you'd need to reattach on every reboot so that portabled can give it the new config. And that might be fine for dbus which is not restartable, but polkit is, so it would just lose it on restart. Unless you mean that the service itself should feed its configuration to dbus/polkit/everything, but that means any unprivileged process could feed anything it wants into dbus/polkit/etc, and that surely cannot work from the security standpoint. And then again, you'd need to change every single component to accept configuration like that - this is unrealistic at best.

**Comment by bluca** [comment]
> And please stop playing games with the needs-discussion label. Dropping it doesnt resolve the fact that this approach is no acceptable in my eyes.

You and I both know that "needs-discussion" doesn't actually mean "needs discussion"

**Comment by poettering** [comment]
> Portabled is a configuration manager, like etcd or ansible. It already writes several things to /etc/ when attaching, that's literally how it works today. Changing the host's mount table under its feet is the opposite of isolation.

etcd is not a configuration manager, it's a distributed key value store.

portabled is not a configuration manage either. it copies one very specific set of files into a separate dir on the host that is clearly separated from everything else and only exists for this purpose. It has an very very narrow focus, and there's clear ownership rules for those resources, and the separate little dir we put this stuff is doubly owned by us (because it's unit files, which are a systemd concept, and then it's a separated dir "system.attached/" that only exists for portabled, and nothing else).

ansible has a much much wider focus, it's a mechanism for managing basically everthing in /etc/.

These schemes are not comparable at all.

**Comment by poettering** [comment]
> > Deeply unimpressed by that. In 2024 we are not going to add additional infra with this as a reason. Just fix your toolchain. I never heard a technical reason why you couldnt fix your toolchain
> 
> I've told you many times, and you don't like the answer - that is fine of course, but you can't use that to tell me to invest a gigantic amount of time and resources to go work on something completely different, that I have literally no use for. Again if someone wants to try and invest time to implement such a scheme, by all means go for it, but it's just not going to be me.

by now you've spent more time arguing about this then it would have taken to just fix the situation and make things proper DDI.

Still, I am not aware of any technical reason why your toolset wouldn't just generate DDIs. This would solve *so* many issues, and make things so much simpler.

> I am really not sure this can be defined as a "corner case", given I have the _very_ strong suspicion I am pretty much the only one in the world that uses any of this for real, in production. I mean either that, or the implementation is absolutely perfect and has zero bugs, and given at this point I've authored a very large chunk of it, I have very serious reservations about the latter option...

I know for a fact that you aren't the only user. Yes, you might have a more substantial usecase than others (i.e. most others I am aware of simply have no interest in providing D-Bus services via portable services), but this doesn't really mean we throw any kind of clean design on the trash.

> No, that doesn't work, and it's not enough. It wouldn't work on reboot for example, you'd need to reattach on every reboot so that portabled can give it the new config. 

Uh? The idea that D-Bus folks mostly all agreed to was to submit the relevant dbus policy at moment of name registration and to hand that off to systemd. The broker would provide a method call that pid 1 could call for creating a new socket pair that is pre-connected and has a name and policy on it. We'd expose this as a new ListenBus= line in .socket units or similar, and then pass the readily set up fd to invoked processes.

**Comment by poettering** [comment]
> You and I both know that "needs-discussion" doesn't actually mean "needs discussion"

it actually means different things for different cases.

**Comment by bluca** [comment]
> > Portabled is a configuration manager, like etcd or ansible. It already writes several things to /etc/ when attaching, that's literally how it works today. Changing the host's mount table under its feet is the opposite of isolation.
> 
> etcd is not a configuration manager, it's a distributed key value store.
> 
> portabled is not a configuration manage either. it copies one very specific set of files into a separate dir on the host that is clearly separated from everything else and only exists for this purpose. It has an very very narrow focus, and there's clear ownership rules for those resources, and the separate little dir we put this stuff is doubly owned by us (because it's unit files, which are a systemd concept, and then it's a separated dir "system.attached/" that only exists for portabled, and nothing else).
> 
> ansible has a much much wider focus, it's a mechanism for managing basically everthing in /etc/.
> 
> These schemes are not comparable at all.

portabled is an asset manager for /etc/ and writes to _multiple_ directories there, and always has. It writes configuration files for a separate service, systemd, that is, these files are created _by_ portabled and are _not used by_ portabled. It also writes to /etc/portabled/, which it owns, and then systemd uses assets from it. portabled does not own /etc/systemd/, just like it doesn't own /etc/dbus-1 or /etc/polkit. The fact that it happens to live in the same git repository is completely orthogonal, because ownership matters at runtime on a system, and there it's a separate and distinct component, the fact that the sources share a repository does not establish any special ownership rules in and by itself.

Asset managers for /etc/ are common and a dime a dozen, and they work on the whole thing or a subset or both, and there is nothing wrong with that. This idea that somehow it's now wrong for an asset manager to write configuration file for another component is completely, 100% wrong. Just look at any orchestrator. A component writing a config file for another component is again how integration between component has always worked and will always work on Linux.

**Comment by bluca** [comment]
> > You and I both know that "needs-discussion" doesn't actually mean "needs discussion"
> 
> it actually means different things for different cases.

Come on now, you tell me not to "play games with labels" but then don't do it yourself. We both know that this label means only one thing: "ignore this"

**Comment by bluca** [comment]
> > They mess up the mount table on the host. Which is fine if that's what one _wants_ to do explicitly, and if everything supports it. It's pretty much only useful for things that are run to completion, like users running binaries or one-shot short-running services. It really doesn't work well for anything else, especially not dynamically. Requiring a reboot to install a new portable service is just not going to work.
> 
> we change the mount table already at any time, for things like logind, sysext/confext, automounts, homed stuff, and so on, it's _ok_ to change mount tables if we do so in a bounded and reasonable way.

Yes, in a bounded and reasonable way. This isn't one. Deploying a separate image that then silently implies as a side effect that the entire composition of the rootfs radically changes, and half of your services see the old version and half the new version, is not bounded nor reasonable.

> logind

Really? How?

> sysext/confext

Sure, when it's explicitly used, that's fine

> automounts

Same as above

> homed

Same as above

**Comment by bluca** [comment]
> > > Deeply unimpressed by that. In 2024 we are not going to add additional infra with this as a reason. Just fix your toolchain. I never heard a technical reason why you couldnt fix your toolchain
> > 
> > 
> > I've told you many times, and you don't like the answer - that is fine of course, but you can't use that to tell me to invest a gigantic amount of time and resources to go work on something completely different, that I have literally no use for. Again if someone wants to try and invest time to implement such a scheme, by all means go for it, but it's just not going to be me.
> 
> by now you've spent more time arguing about this then it would have taken to just fix the situation and make things proper DDI.
> 
> Still, I am not aware of any technical reason why your toolset wouldn't just generate DDIs. This would solve _so_ many issues, and make things so much simpler.

I have already mentioned it many times: dealing with GPT partitions sucks, both when creating them and when using them, because it takes time to probe and whatnot, and it's all wasted and pointless use of resources in the critical path. A single filesystem is just that, it can be just used straight away, no need for waiting for the kernel, udev or anything else. This makes a huge difference in the hot path.
Also, it is now an API, and it cannot be "just fixed", everything relies on having a single filesystem image and the 3 verity files, and cannot change anymore, as everything would break once the format changes and some files disappear, just like you can't just make incompatible changes in any other public API that is actively used. And no, saying "just fix it" is not helpful in any way, shape or form.

> > I am really not sure this can be defined as a "corner case", given I have the _very_ strong suspicion I am pretty much the only one in the world that uses any of this for real, in production. I mean either that, or the implementation is absolutely perfect and has zero bugs, and given at this point I've authored a very large chunk of it, I have very serious reservations about the latter option...
> 
> I know for a fact that you aren't the only user. Yes, you might have a more substantial usecase than others (i.e. most others I am aware of simply have no interest in providing D-Bus services via portable services), but this doesn't really mean we throw any kind of clean design on the trash.

I mean, where are they? In the past 3 years there are literally zero contributions from anyone but me excluding tree-wide cleanups/changes, and I do not remember a single bug report either. I have seriously a hard time believing everything just works flawlessly...
Also we literally had a call last Friday with someone who looked a this, and then discarded it because it couldn't do polkit/dbus integration. It was the wrong tool for that job anyway, but still, it happened.

> > No, that doesn't work, and it's not enough. It wouldn't work on reboot for example, you'd need to reattach on every reboot so that portabled can give it the new config.
> 
> Uh? The idea that D-Bus folks mostly all agreed to was to submit the relevant dbus policy at moment of name registration and to hand that off to systemd. The broker would provide a method call that pid 1 could call for creating a new socket pair that is pre-connected and has a name and policy on it. We'd expose this as a new ListenBus= line in .socket units or similar, and then pass the readily set up fd to invoked processes.

So in other words, extracting and writing out config files in a component that extracts and writes out config files is bad because:

> let's not do this with explicit hookups for some select packages. That doesn't scale.

...but adding support directly in PID1 for all configuration formats for any possible third party service, and adding a special handover protocol that requires changes in systemd, the service code and the third party component code, and a special unit setting for pointing to said configuration in order to parse it (which is XML on top of that, I thought you never wanted to deal with parsing XML in systemd?), is somehow not an explicit hookup and scales well?

**Comment by poettering** [comment]
> Come on now, you tell me not to "play games with labels" but then don't do it yourself. We both know that this label means only one thing: "ignore this"

That's nonsense. It means different things in different contexts. The common theme though is: a reviewer as serious doubts about this.

**Comment by poettering** [comment]
> Really? How?

every user as they log in gets a private XDG_RUNTIME_DIR mounted

**Comment by poettering** [comment]
> I have already mentioned it many times: dealing with GPT partitions sucks, both when creating them and when using them, because it takes time to probe and whatnot, and it's all wasted and pointless use of resources in the critical path. 

it's a trivial data structure around a file system, that's all. it takes no measurable resources to parse, and it's widely undestood by tools, libraries and the kernel. And it's everything but pointless, since it allows treating stuff that belongs together as one.

> A single filesystem is just that, it can be just used straight away, no need for waiting for the kernel, udev or anything else. 

Uh? It's not a single file system though, it's a bunch of files, that need to be downloaded, copied, updated, maintained, authenticated together, and that get out of sync, it's a frickin mess. And we do not wait for kernel/udev anymore for parsing GPT, we haven#t for a long long time. We create the partitions we need via ioctls, and immediately use them. It's synchronous, immediate, trivial.

yes, activating disk images is slow, but that's not the fault of DDIs, but of dm-verity and it's messy interaction with udev.

> This makes a huge difference in the hot path.

It certainly does not. we issue one extra ioctl per partition we care about. that's hardly measurable. 

> Also, it is now an API, and it cannot be "just fixed", everything relies on having a single filesystem image and the 3 verity files, and cannot change anymore, as everything would break once the format changes and some files disappear, just like you can't just make incompatible changes in any other public API that is actively used. And no, saying "just fix it" is not helpful in any way, shape or form.

you do realize that this comes up again and again and again. You are just piling up technical debt.

**Comment by YHNdnzj** [comment]
Yeah, tech debt from downstream is really not a reason to introduce narrowly-focused stuff upstream. If something ought to be upstreamed it should be reasonably generic and extendable. GPT images are superior on this, and we have all the infrastructures in the codebase, hence focusing on one use case/a small and deliberately-selected set of softwares would only mean more tech debt in the future.

**Comment by bluca** [comment]
> > Really? How?
> 
> every user as they log in gets a private XDG_RUNTIME_DIR mounted

That doesn't change the rootfs? It's a subdirectory in /run/user/ which is in another tmpfs already, not quite the same

**Comment by bluca** [comment]
> > I have already mentioned it many times: dealing with GPT partitions sucks, both when creating them and when using them, because it takes time to probe and whatnot, and it's all wasted and pointless use of resources in the critical path.
> 
> it's a trivial data structure around a file system, that's all. it takes no measurable resources to parse, and it's widely undestood by tools, libraries and the kernel. And it's everything but pointless, since it allows treating stuff that belongs together as one.

It's really not widely understood, not as much as a simple squashfs anyway, in reality. For example, you can trivially build a squashfs file on Windows or OSX. I haven't the faintest clue if you can do a complex GPT images with linux filesystems inside there, but for sure none of our tools that make dealing with GPT images bearable work on Windows or OSX - you can't really run sd-repart there.

> > A single filesystem is just that, it can be just used straight away, no need for waiting for the kernel, udev or anything else.
> 
> Uh? It's not a single file system though, it's a bunch of files, that need to be downloaded, copied, updated, maintained, authenticated together, and that get out of sync, it's a frickin mess. And we do not wait for kernel/udev anymore for parsing GPT, we haven#t for a long long time. We create the partitions we need via ioctls, and immediately use them. It's synchronous, immediate, trivial.

Moving files around is trivial and it's never been a problem, not sure where you got that impression, but it's definitely not my experience in 5 years of using this across a gazillion systems. And yes, you still wait for the kernel, as the ioctls are kernel calls. It is not immediate, and it is not trivial.

> yes, activating disk images is slow, but that's not the fault of DDIs, but of dm-verity and it's messy interaction with udev.

Nope, that's the same across both setups after all. The difference is the partition table and all that comes from it.

> > This makes a huge difference in the hot path.
> 
> It certainly does not. we issue one extra ioctl per partition we care about. that's hardly measurable.

It very much did last time I checked. I mean maybe not on an idle last-gen top-line x86_64 cpu, but the world doesn't solely consist of those, some of us are stuck on low power hardware

> > Also, it is now an API, and it cannot be "just fixed", everything relies on having a single filesystem image and the 3 verity files, and cannot change anymore, as everything would break once the format changes and some files disappear, just like you can't just make incompatible changes in any other public API that is actively used. And no, saying "just fix it" is not helpful in any way, shape or form.
> 
> you do realize that this comes up again and again and again. You are just piling up technical debt.

It's not technical debt

**Comment by bluca** [comment]
> Yeah, tech debt from downstream is really not a reason to introduce narrowly-focused stuff upstream. If something ought to be upstreamed it should be reasonably generic and extendable. GPT images are superior on this, and we have all the infrastructures in the codebase, hence that should be used. Focusing on one use case/a small and deliberately-selected set of softwares would only mean more tech debt in the future.

It's not technical debt, it's working exactly as it is meant to. Just because somebody doesn't like something, it doesn't make it technical debt. In fact, simply in terms of deployments, I am pretty sure it is order of magnitudes more used than GPT images. Let's be realistic here, apart from us running mkosi for tests and local development, and a few other folks running experiments here and there, I don't think anybody else is actually using these...

**Comment by bluca** [comment]
> > > No, that doesn't work, and it's not enough. It wouldn't work on reboot for example, you'd need to reattach on every reboot so that portabled can give it the new config.
> > 
> > 
> > Uh? The idea that D-Bus folks mostly all agreed to was to submit the relevant dbus policy at moment of name registration and to hand that off to systemd. The broker would provide a method call that pid 1 could call for creating a new socket pair that is pre-connected and has a name and policy on it. We'd expose this as a new ListenBus= line in .socket units or similar, and then pass the readily set up fd to invoked processes.
> 
> So in other words, extracting and writing out config files in a component that extracts and writes out config files is bad because:
> 
> > let's not do this with explicit hookups for some select packages. That doesn't scale.
> 
> ...but adding support directly in PID1 for all configuration formats for any possible third party service, and adding a special handover protocol that requires changes in systemd, the service code and the third party component code, and a special unit setting for pointing to said configuration in order to parse it (which is XML on top of that, I thought you never wanted to deal with parsing XML in systemd?), is somehow not an explicit hookup and scales well?

I mean it's funny because now suddenly it's really bad and poorly thought through and technical debt and terrible and so on and so forth if systemd components install config files that other select components use, for some reasons. But when you wanted to use ssh with some particular custom configurations for some reason that wasn't a problem, so systemd happily installs drop-in configuration that only openssh client/server understand in /etc/ssh/:

https://github.com/systemd/systemd/commit/7e3607996a5df22799dd092526c78b5bd7e245f4
https://github.com/systemd/systemd/commit/0abd510f7f628d0369f0814b671302e93c62b161

This isn't even a generic ssh thing, it's only for the individual openssh implementation, and doesn't work for dropbear and other implementations. It's exactly the same approach of this PR - worse even, as this PR works with dbus-broker and dbus-daemon and zbus, so it's implementation-agnostic.

So I guess we should revert those two commits, and you should instead go implement an IPC protocol in openssh to take configuration in from units when they are started instead? I mean the same standard either applies to everything or to nothing, surely?

**Comment by bluca** [comment]
> > I mean it's funny because now suddenly it's really bad and poorly thought through and technical debt and terrible and so on and so forth if systemd components install config files that other select components use, for some reasons. But when you wanted to use ssh with some particular custom configurations for some reason that wasn't a problem, so systemd happily installs drop-in configuration that only openssh client/server understand in /etc/ssh/
> 
> Please, let's not confuse completely different concepts. We install _our_ configs to sshd/dbus/polkit dirs, in order to integrate _our code_. We control which components we need to do integration with, and that's the correct way to do things. But here the images are provided by users, and it's wrong to selectively copy some files to the host.
> 
> IOW, sshd/dbus/polkit are special to us because we use them. But how are these components special to user-provided images? Users should control which files to integrate, not us.

It's the same thing. The objection, literal quoting, was:

> let's not mix and match stuff in dirs owned by other packages.

openssh owns those directories - not even a generic ssh protocol, but the specific openssh implementation, carefully hand picked and selected. Now you are further moving the goal posts, and it's obvious why: you use the openssh integration so it's convenient to have it and you know that this config-via-IPC stuff is never going to go anywhere, but you don't care about any of this, so this unjustified and novel purity of directory ownership, not found anywhere else, suddenly starts to matter.

Portabled does not own _any_ of the resources that are extracted. And yet it still extracts them, as it should: it's an asset manager, its entire purpose is to attach images and extract the configuration files needed to integrate in the system, update them, and remove them. It doesn't own _any_ of the configuration files it currently extracts, like any other asset manager/orchestrator/deployer/package managers/whatever, the purpose of these tools is exclusively to bridge component A with component B/C/D, and the normal way to do that on Linux is to add drop-in configuration files. Assets managers like apt/dpkg do not let the users choose what to integrate either - they have specifically defined hooks for specifically defined components, and if the payload contains them, they are integrated. Same thing for Flatpak, and Snaps, and a billion other things. This is the normal and standard approach across a myriad of different parts of the Linux ecosystem.
